packages/edit-site/src/components/sidebar-navigation-screen-pages/index.js

```diff
@@ -5,64 +5,157 @@ import {
-	const { records: pages, isResolving: isLoading } = useEntityRecords(
+	const { records: pages, isResolving: isLoadingPages } = useEntityRecords(
 		'postType',
 		'page',
-		{ status: 'any' }
+		{
+			status: 'any',
+			per_page: -1,
+		}
 	);
+	const { records: templates, isResolving: isLoadingTemplates } =
+		useEntityRecords( 'postType', 'wp_template', {
+			per_page: -1,
+		} );
+
+	const dynamicPageTemplates = templates?.filter( ( { slug } ) =>
+		[ '404', 'search' ].includes( slug )
+	);
+
+	const homeTemplate =
+		templates?.find( ( template ) => template.slug === 'front-page' ) ||
+		templates?.find( ( template ) => template.slug === 'home' ) ||
+		templates?.find( ( template ) => template.slug === 'index' );
+
+	const pagesAndTemplates = pages?.concat( dynamicPageTemplates, [
+		homeTemplate,
+	] );
+
+	const { frontPage, postsPage } = useSelect( ( select ) => {
+		const { getEntityRecord } = select( coreStore );
+
+		const siteSettings = getEntityRecord( 'root', 'site' );
+		return {
+			frontPage: siteSettings?.page_on_front,
+			postsPage: siteSettings?.page_for_posts,
+		};
+	}, [] );
+
+	const isHomePageBlog = frontPage === postsPage;
+
+	const reorderedPages = pages && [ ...pages ];
+
+	if ( ! isHomePageBlog && reorderedPages?.length ) {
+		const homePageIndex = reorderedPages.findIndex(
+			( item ) => item.id === frontPage
+		);
+		const homePage = reorderedPages.splice( homePageIndex, 1 );
+		reorderedPages?.splice( 0, 0, ...homePage );
+
+		const postsPageIndex = reorderedPages.findIndex(
+			( item ) => item.id === postsPage
+		);
+
+		const blogPage = reorderedPages.splice( postsPageIndex, 1 );
+
+		reorderedPages.splice( 1, 0, ...blogPage );
+	}
 
 	return (
 		<SidebarNavigationScreen
 			title={ __( 'Pages' ) }
 			description={ __( 'Browse and edit pages on your site.' ) }
 			content={
 				<>
-					{ isLoading && (
+					{ ( isLoadingPages || isLoadingTemplates ) && (
 						<ItemGroup>
 							<Item>{ __( 'Loading pages' ) }</Item>
 						</ItemGroup>
 					) }
-					{ ! isLoading && (
-						<>
-							<SidebarNavigationSubtitle>
-								{ __( 'Recent' ) }
-							</SidebarNavigationSubtitle>
-							<ItemGroup>
-								{ ! pages?.length && (
-									<Item>{ __( 'No page found' ) }</Item>
-								) }
-								{ pages?.map( ( page ) => (
+					{ ! ( isLoadingPages || isLoadingTemplates ) && (
+						<ItemGroup>
+							{ ! pagesAndTemplates?.length && (
+								<Item>{ __( 'No page found' ) }</Item>
+							) }
+							{ isHomePageBlog && homeTemplate && (
+								<PageItem
+									postType="wp_template"
+									postId={ homeTemplate.id }
+									key={ homeTemplate.id }
+									icon={ home }
+									withChevron
+								>
+									<Truncate numberOfLines={ 1 }>
+										{ decodeEntities(
+											homeTemplate.title?.rendered
+										) ?? __( '(no title)' ) }
+									</Truncate>
+								</PageItem>
+							) }
+							{ reorderedPages?.map( ( item ) => {
```

Comment:
Optional, but perhaps this could be simplified with destructuring, and a function to grab the icon.

```jsx

	const getIcon = ( id ) => { // I don't think we need useCallback here since `frontPage` and `postsPage` won't change after defined.
		if ( id === frontPage ) {
			return home;
		}
		if ( id === postsPage ) {
			return loop;
		}
		return page;
	};

....


{ dynamicPageTemplates?.map(
	( { id, title } ) => (
		<PageItem
			postType="wp_template"
			postId={ id }
			key={ id }
			icon={ layout }
			withChevron
		>
			<Truncate numberOfLines={ 1 }>
				{ decodeEntities(
					title?.rendered
				) ?? __( '(no title)' ) }
			</Truncate>
		</PageItem>
	)
) }
```